### PR TITLE
Fix #11352 strutil.insertSep() fails on negative numbers

### LINKED ISSUE #11352: strutil.insertSep() fails on negative numbers
Function `insertSep` outputs the wrong string.

### Example

```nim
insertSep($(-100), ',')
insertSep($(-100_000), ',')
```

### Current Output
```
-,100
-,100,000
```

### Expected Output

```
-100
-100,000
```

### Possible Solution

A couple options come to mind.  Don't know which fits better with the library, but placing separators in `ints` and `floats` is something done very often.

* Put in special case code in `insertSep` to handle the case when the first character is a minus sign.
* Create an overloaded proc of `insertSep(s: int): string` that handles the special case of negative numbers

### Additional Information

```
$ nim -v
Nim Compiler Version 0.19.9 [MacOSX: amd64]
Compiled at 2019-05-21
```

### PR DESCRIPTION
**Pull request by bung87**
https://github.com/nim-lang/Nim/issues/11352

### CODE REVIEW

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -2264,19 +2264,34 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
...
+    prefix.add s[s.low]
+    for i in (s.low + 1)..s.high:
+      idx = i
+      if not isDigit(s[i]):
+        prefix.add s[i]
+      else:
+        break
+    parts = s.substr(idx)
+  else:
+    parts.shallowCopy s
```
Comment (alaviss) [inline lib/pure/strutils.nim]: Please avoid `shallowCopy`, it's semantic is confusing and ARC will feature move semantics which is superior instead.

File: lib/pure/strutils.nim [hunk 2]
```diff
@@ -2264,19 +2264,34 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
...
+  if hasPrefix:
+    var idx:int
+    prefix.add s[s.low]
+    for i in (s.low + 1)..s.high:
+      idx = i
+      if not isDigit(s[i]):
+        prefix.add s[i]
+      else:
+        break
+    parts = s.substr(idx)
```
Comment (Araq) [inline lib/pure/strutils.nim]: Slow.

File: lib/pure/strutils.nim [hunk 3]
```diff
@@ -2264,19 +2264,34 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
...
+  else:
+    parts = s
```
Comment (Araq) [inline lib/pure/strutils.nim]: Slow.

File: lib/pure/strutils.nim [hunk 4]
```diff
@@ -2264,19 +2264,34 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
...
+  let partsLen = parts.len
```
Comment (Araq) [inline lib/pure/strutils.nim]:
```nim

  let partsLen = s.len - idx

```

File: lib/pure/strutils.nim [hunk 5]
```diff
@@ -2264,19 +2264,34 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
...
   var j = 0
   dec(L)
-  for i in countdown(len(s)-1, 0):
+  for i in countdown(partsLen-1, 0):
     if j == digits:
       result[L] = sep
       dec(L)
       j = 0
-    result[L] = s[i]
+    result[L] = parts[i]
```
Comment (Araq) [inline lib/pure/strutils.nim]:
```nim

result[L] = s[i + idx]

```

File: lib/pure/strutils.nim [hunk 6]
```diff
@@ -2264,19 +2264,34 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
...
     inc(j)
     dec(L)
+  result = prefix & result
```
Comment (Araq) [inline lib/pure/strutils.nim]: Slow. Coding this without the copies is easy enough really.

File: lib/pure/strutils.nim [hunk 7]
```diff
@@ -2264,17 +2264,27 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
...
+      if not isDigit(s[i]):
+        result.add s[i]
+      else:
+        break
+  let partsLen = s.len - idx
+  var L = (partsLen-1) div digits + partsLen
   var j = 0
   dec(L)
-  for i in countdown(len(s)-1, 0):
+  for i in countup(0, partsLen-1):
```
Comment (alaviss) [inline lib/pure/strutils.nim]:
```suggestion
  for i in 0 ..< partsLen:
```

**Pull request merged**